### PR TITLE
feat: topolvm storage class resource unit implementation

### PR DESCRIPTION
**Pull request by leelavg**
- Implementation of resourceManager interface for topolvm storage class
- Addition of RBAC for operator to manage topolvm storage classes
- Envtest to verify creation and deletion of storage class based on CR

### CODE REVIEW

**Comment by leelavg** [comment]
**Pending: It'll be part of separate PR**
- Based on lvmCluster CR, storage class unit also extracts tolerations
- It'll check the status of daemonset with these tolerations and if they are running then volume group creation was successful
- Then storageclass will respect reconciler `ensureCreated` and creates StorageClass or else it'll wait for next cycle.

File: controllers/topolvm_storageclass.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+				Annotations: map[string]string{
+					"description": "Provides RWO and RWOP Filesystem & Block volumes",
+				},
+			},
+			Provisioner:          TopolvmCSIDriverName,
+			VolumeBindingMode:    &volumeBindingMode,
+			AllowVolumeExpansion: &allowVolumeExpansion,
+			Parameters: map[string]string{
+				fmt.Sprintf("%s/device-class", TopolvmCSIDriverName): deviceClass.Name,
+				"csi.storage.k8s.io/fstype":                          "xfs",
```
Comment (nbalacha) [inline controllers/topolvm_storageclass.go:104]: Minor comment .Please move xfs to a default var.

File: controllers/topolvm_storageclass.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	// delete the corresponding storage class
+	for _, deviceClass := range lvmCluster.Spec.DeviceClasses {
+		sc := &storagev1.StorageClass{}
+		scName := fmt.Sprintf("topolvm-%s", deviceClass.Name)
+		err := r.Client.Get(ctx, types.NamespacedName{Name: scName}, sc)
+
+		if err != nil {
+			// already deleted in previous reconcile
+			if errors.IsNotFound(err) {
+				r.Log.Info("topolvm storage class", "StorageClass", scName)
```
Comment (jmolmo) [inline controllers/topolvm_storageclass.go:60]: Message is not clear... better something like:
"topolvm storage class already deleted"

File: controllers/topolvm_storageclass.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+			r.Log.Error(err, "failed to retrieve topolvm storage class", "StorageClass", scName)
+			return err
+		}
+
+		// storageClass exists, initiate deletion
+		if sc.GetDeletionTimestamp().IsZero() {
+			if err = r.Client.Delete(ctx, sc); err != nil {
+				r.Log.Error(err, "failed to delete topolvm storage class", "StorageClass", scName)
+				return err
+			}
```
[the remark this answers is not in the document]
Comment (leelavg) [reply controllers/topolvm_storageclass.go:72]: above means we are unable to delete storage class or did I miss anything?

File: controllers/topolvm_storageclass.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+func getTopolvmStorageClasses(lvmCluster *lvmv1alpha1.LVMCluster) []*storagev1.StorageClass {
+	sc := []*storagev1.StorageClass{}
+	allowVolumeExpansion := true
+	volumeBindingMode := storagev1.VolumeBindingWaitForFirstConsumer
+
+	for _, deviceClass := range lvmCluster.Spec.DeviceClasses {
+		storageClass := &storagev1.StorageClass{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("topolvm-%s", deviceClass.Name),
```
Comment (sp98) [inline controllers/topolvm_storageclass.go:94]: would this ensure correct [naming conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names)? Can device class have upper case alphanumeric chars?

File: controllers/topolvm_storageclass.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmolmo) [reply controllers/topolvm_storageclass.go:72]: I mean.. just inform when the sc is deleted: 
```
		if sc.GetDeletionTimestamp().IsZero() {
			if err = r.Client.Delete(ctx, sc); err != nil {
				r.Log.Error(err, "failed to delete topolvm storage class", "StorageClass", scName)
				return err
			} else {
				r.Log.Info(err, "topolvm storage class succesfully deleted", "StorageClass", scName)
			}
```

File: controllers/topolvm_storageclass.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+				Annotations: map[string]string{
+					"description": "Provides RWO and RWOP Filesystem & Block volumes",
```
Comment (sp98) [inline controllers/topolvm_storageclass.go:96]: are we supporting RWOP?

File: config/rbac/role.yaml [hunk 6]
```diff
@@ -65,3 +65,13 @@ rules:
...
+- apiGroups:
+  - storage.k8s.io
+  resources:
+  - storageclasses
+  verbs:
+  - create
+  - delete
+  - get
+  - list
+  - watch
```
Comment (sp98) [inline config/rbac/role.yaml:77]: where do we need `watch`?

File: controllers/topolvm_storageclass.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (leelavg) [reply controllers/topolvm_storageclass.go:72]: it's already logged above and this codepath will not be touched in that case

**Review by jmolmo — CHANGES_REQUESTED** [review]
I think it is OK. LGTM after make just a couple of changes in the log messages. I would like to comment that the name of the resource is "singular"  topolvmStorageClass, but what we are managing are several storageclasses. probably a name like "topolvmStorageClasses" would be more meaningful, but it is just an opinion. It is is ok to keep the name in singular.

File: config/rbac/role.yaml (hunk 6 above)
Comment (leelavg) [reply config/rbac/role.yaml:77]: - it's probably for operator-sdk/controller-runtime to build sharedinformercache, I didn't dig into code though where that was defined
- when `watch` wasn't provided I saw errors logged in

File: controllers/topolvm_storageclass.go (hunk 5 above)
Comment (leelavg) [reply controllers/topolvm_storageclass.go:96]: - either ways, there isn't any watch to limit access mode and I thought it's better to annotate the capabilities that the provisioner provides

File: controllers/topolvm_storageclass.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmolmo) [reply controllers/topolvm_storageclass.go:72]: you mean in a second pass?

[the remark this answers is not in the document]
Comment (leelavg) [reply controllers/topolvm_storageclass.go:72]: ack, I think it doesn't harm to log deletion is initiated.

File: controllers/topolvm_storageclass.go (hunk 4 above)
Comment (leelavg) [reply controllers/topolvm_storageclass.go:94]: - good catch, do not want to implement it here, will track this as an issue. Sounds good?
- it's better to restrict deviceClass to lowercase but let's think about that later.

Comment (sp98) [reply controllers/topolvm_storageclass.go:96]: sounds good to me.

**Review by sp98 — APPROVED** [review]

**Review by jmolmo — APPROVED** [review]

**Review by nbalacha — APPROVED** [review]

**Pull request merged**